### PR TITLE
Fix MouseTracker annotation leak

### PR DESCRIPTION
**Pull request by ds84182**
## Description

MouseTracker had a bug where `Map.remove` was being called with a value instead of a key, causing widgets and their state to leak (if it used a Listener).

Changed the existing pointer listener test to check if the annotation has been removed from the MouseTracker.

## Related Issues

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by goderbauer** [comment]
@ds84182 Thanks for your contribution. It's good practice to make only one change per PR. That way, we can easily roll a specific change back if it causes trouble.

Also, it looks like you change is causing some existing tests to fail. Did you look at those failures? 

/cc @gspencergoog

**Comment by ds84182** [comment]
Gotcha, I'll move the double painting to another PR. Seems like the issue comes from marking RenderPointerListener as always composited.

**Comment by ds84182** [comment]
Update, PTAL

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 1]
```diff
@@ -124,7 +124,7 @@ class MouseTracker {
     for (int deviceId in trackedAnnotation.activeDevices) {
       annotation.onExit(PointerExitEvent.fromHoverEvent(_lastMouseEvent[deviceId]));
     }
-    _trackedAnnotations.remove(trackedAnnotation);
+    _trackedAnnotations.remove(annotation);
```
Comment (gspencergoog) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: Nice find, thanks!

File: packages/flutter/test/widgets/listener_test.dart [hunk 2]
```diff
@@ -125,6 +129,7 @@ void main() {
       ));
       expect(exit, isNotNull);
       expect(exit.position, equals(const Offset(400.0, 300.0)));
+      expect(WidgetsBinding.instance.mouseTracker.isAnnotationAttached(renderListener.hoverAnnotation), isFalse);
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/listener_test.dart]: Instead, you can use `tester.binding.mouseTracker`...
```suggestion
      expect(tester.binding.mouseTracker.isAnnotationAttached(renderListener.hoverAnnotation), isFalse);
```

File: packages/flutter/test/widgets/listener_test.dart [hunk 3]
```diff
@@ -109,6 +112,7 @@ void main() {
           onPointerExit: (PointerExitEvent details) => exit = details,
         ),
       ));
+      final RenderPointerListener renderListener = listenerKey.currentContext.findRenderObject();
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/listener_test.dart]: Here, you can use a finder:
```suggestion
      final RenderPointerListener renderListener = tester.renderObject(find.byType(Listener));
```
That way you don't need the GlobalKey.

File: packages/flutter/lib/src/gestures/mouse_tracking.dart [hunk 4]
```diff
@@ -171,6 +171,12 @@ class MouseTracker {
     return trackedAnnotation;
   }
 
+  /// Checks if the given [MouseTrackerAnnotation] is attached to this
+  /// [MouseTracker].
+  bool isAnnotationAttached(MouseTrackerAnnotation annotation) {
```
Comment (gspencergoog) [inline packages/flutter/lib/src/gestures/mouse_tracking.dart]: Make this `@visibleForTesting` too: it's not generally useful, so we don't want to add to the API surface unnecessarily.

**Review by gspencergoog — CHANGES_REQUESTED** [review]

**Comment by ds84182** [comment]
Should be good to go now

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by gspencergoog** [comment]
I'll land this as soon as the tests are done running.

Thanks so much for the PR!

**Comment by gspencergoog** [comment]
Thanks for merging that, @dnfield! I just forgot to check back on this to see that it had finished.

**Pull request merged**